nncf/common/pruning/schedulers.py

```diff
@@ -10,112 +10,146 @@
  See the License for the specific language governing permissions and
  limitations under the License.
 """
+
+from typing import Optional
 import numpy as np
 import scipy.optimize
 
 from nncf.api.compression import CompressionScheduler
-from nncf.config import NNCFConfig
 from nncf.common.utils.registry import Registry
+from nncf.common.schedulers import ExponentialDecaySchedule
 
 PRUNING_SCHEDULERS = Registry("pruning_schedulers")
 
 
 class PruningScheduler(CompressionScheduler):
-    def __init__(self, pruning_algo, params: NNCFConfig = None):
-        super().__init__()
-        if params is None:
-            self._params = NNCFConfig()
-        else:
-            self._params = params
+    """
+    This is the class from which all pruning schedulers inherit.
 
-        self.algo = pruning_algo
+    A pruning scheduler is an object which specifies the pruning
+    level at each training epoch. It involves a scheduling algorithm,
+    defined in the `_calculate_pruning_level()` method and a state
+    (some parameters required for current pruning level calculation)
+    defined in the `__init__()` method.
+    """
 
-        # Number of initial steps of training before pruning
-        self.num_init_steps = self._params.get('num_init_steps', 0)
-        self.pruning_steps = self._params.get('pruning_steps', 100)
+    def __init__(self, controller, params: dict):
+        """
+        Initializes the internal state of the pruning scheduler specified by:
+            - controller: Pruning algorithm controller.
+            - initial_pruning: Pruning level which already has been
+                applied to the model. It is the level at which the schedule begins.
+            - target_pruning: Pruning level at which the schedule ends.
+            - num_warmup_epochs: Number of epochs for model pre-training before pruning.
+            - num_pruning_epochs: Number of epochs during which the pruning level
+                is increased from `initial_pruning` to `target_pruning`.
+            - freeze_epoch: Zero-based index of the epoch from which the pruning
+                mask will be frozen and will not be trained.
+            - current_pruning: Pruning level for the `current_epoch`.
+
+        :param controller: Pruning algorithm controller.
+        :param params: Parameters of the scheduler.
+        """
+        super().__init__()
+        self.controller = controller
+        self.initial_pruning = self.controller.pruning_init
 
-        # Pruning rates
-        self.initial_pruning = self.algo.pruning_init
-        if self.algo.prune_flops:
-            self.pruning_target = self._params.get('pruning_flops_target')
+        if self.controller.prune_flops:
+            self.target_pruning = params.get('pruning_flops_target')
         else:
-            self.pruning_target = self._params.get('pruning_target', 0.5)
+            self.target_pruning = params.get('pruning_target', 0.5)
```

Comment:
Does not look well. Now the name of the parameter in the config and the name of the attribute differ. This is not critical, but affects debuggability.

Reply:
I can rename
`target_pruning` --> `pruning_target`
`initial_pruning` --> `pruning_initial` (for consistency)
and do the sample for sparsity scheduler if needed.

But, I think, that some config parameters don't have a very good name. For example, `num_init_steps` and `pruning_steps`
(`num_warmup_epochs` and `num_pruning_epochs` are better for me).

Reply:
I think you can do the corresponding renamings for `num_init_steps` and `pruning_steps` as well.

Reply:
AR @andrey-churkin to rename the entities in configs in a separate PR

Reply:
After discussions, we decided to use `initial_level`, `target_level`, and `current_level` as attribute names for schedulers.